File: src/main/java/korzin/io/algorithms/BalancedBrackets.java
package korzin.io.algorithms;

import korzin.io.structures.Stack;

import java.util.*;

public class BalancedBrackets {

    private Map<String, String> closeToOpenSymbolsMap = new HashMap<String, String>() {{
        put(")", "(");
        put("]", "[");
    }};

    private Set<String> symbolsToHandle = new HashSet<>(Arrays.asList("(", ")", "[", "]"));

    public boolean hasBalancedBrackets(String strToCheck) {
        String[] sequence = strToCheck.split("");
        Stack<String> stack = new Stack<>();
        for (String s : sequence) {
            if (symbolsToHandle.contains(s)) {
                String head = stack.top();
                String closingSymbol;
                if (head == null
                        || (closingSymbol = closeToOpenSymbolsMap.get(s)) == null
                        || !head.equals(closingSymbol)) {
                    stack.push(s);
                } else {
                    stack.pop();
                }
            }
        }
        return stack.top() == null;
    }
}

File: src/main/java/korzin/io/algorithms/BalancedBrackets.kt
package korzin.io.algorithms

import korzin.io.structures.Stack

class BalancedBrackets {

    private val bracketsOpenCloseMap: Map<String, String> = hashMapOf(")" to "(", "]" to "[")
    private val bracketsSet: Set<String> = setOf("(", ")", "[", "]")

    fun isBracketBalanced(input: String): Boolean {
        val inputLetters: List<String> = input.split("")
        val procStack = Stack<String>()
        for (letter: String in inputLetters) {
            if (bracketsSet.contains(letter)) {
                if (bracketsOpenCloseMap.containsKey(letter)) {
                    if (procStack.top() == null || procStack.pop() != bracketsOpenCloseMap[letter]) {
                        return false
                    }
                } else {
                    procStack.push(letter)
                }
            }
        }
        return procStack.top() == null;
    }
}
